codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -140,6 +143,100 @@ public void onPriorityTreeParentChanging(Http2Stream stream, Http2Stream newPare
                 }
             }
         });
+
+        if (listener == null) {
+            writeAllocatedBytesVisitor = new RemoteStreamVisitor() {
+                @Override
+                public boolean visit(Http2Stream stream) {
+                    state(stream).writeAllocatedBytes();
+                    return true;
+                }
+
+                @Override
+                public void beforeIteration() {
+                }
+
+                @Override
+                public Http2Stream streamMadeConnectionNotWritable() {
+                    return null;
+                }
+            };
+            incrementWindowSizeWritabilityVisitor = initialWindowSizeWritabilityVisitor =
+                    writePendingBytesWritabilityVisitor = null;
+        } else {
+            writeAllocatedBytesVisitor = new RemoteStreamVisitor() {
+                private boolean connectionOriginalWritable;
+                private Http2Stream streamMadeConnectionNotWritable;
+                @Override
+                public boolean visit(Http2Stream stream) {
+                    AbstractState state = state(stream);
+                    int written = state.writeAllocatedBytes();
```

Comment:
> Maybe we should skip writing if streamMadeConnectionNotWritable != null? Then maybe we could get rid of writePendingBytesWritabilityVisitor and just make another pass using this visitor to change writability for all streams in the collection prior to streamMadeConnectionNotWritable?

I'm assuming you mean re-use `writeAllocatedBytesVisitor`? Yes we could do something like this. That would require 2 full iterations over the active streams. The current approach has 1 full iteration to do the writing and then 1 iteration up to the element that caused the connection window to be exhausted.